File: examples/go/servers/mcp/existing-server.go
package main

// MCP Server with x402 Paid Tools - Existing Server Integration
//
// This example demonstrates adding x402 to an existing MCP server using NewPaymentWrapper.
// Use this approach when you have an EXISTING MCP server and want to add
// x402 payment to specific tools without adopting the full x402MCPServer abstraction.
//
// The getWeatherData helper is defined in helpers.go and shared across examples.
//
// Run with: go run . existing

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"os"

	"github.com/joho/godotenv"
	mcpsdk "github.com/modelcontextprotocol/go-sdk/mcp"
	x402 "github.com/x402-foundation/x402/go"
	x402http "github.com/x402-foundation/x402/go/http"
	"github.com/x402-foundation/x402/go/mcp"
	evm "github.com/x402-foundation/x402/go/mechanisms/evm/exact/server"
)

func runExisting() error {
	fmt.Println("\n📦 Using NewPaymentWrapper with existing server\n")

	// Load environment variables
	if err := godotenv.Load(); err != nil {
		fmt.Println("No .env file found, using environment variables")
	}

	evmAddress := os.Getenv("EVM_ADDRESS")
	if evmAddress == "" {
		return fmt.Errorf("EVM_ADDRESS environment variable is required")
	}

	facilitatorURL := os.Getenv("FACILITATOR_URL")
	if facilitatorURL == "" {
		return fmt.Errorf("FACILITATOR_URL environment variable is required")
	}

	port := os.Getenv("PORT")
	if port == "" {
		port = "4022"
	}

	// ========================================================================
	// STEP 1: Your existing MCP server (this might already exist in your code)
	// ========================================================================
	mcpServer := mcpsdk.NewServer(&mcpsdk.Implementation{
		Name:    "x402 MCP Server (Existing)",
		Version: "1.0.0",
	}, nil)

	// ========================================================================
	// STEP 2: Set up x402 resource server for payment handling
	// ========================================================================
	ctx := context.Background()
	facilitatorClient := x402http.NewHTTPFacilitatorClient(&x402http.FacilitatorConfig{
		URL: facilitatorURL,
	})

	resourceServer := x402.Newx402ResourceServer(
		x402.WithFacilitatorClient(facilitatorClient),
	)
	resourceServer.Register("eip155:84532", evm.NewExactEvmScheme())

	if err := resourceServer.Initialize(ctx); err != nil {
		return fmt.Errorf("failed to initialize resource server: %w", err)
	}

	// ========================================================================
	// STEP 3: Build payment requirements for different tools
	// ========================================================================
	weatherConfig := x402.ResourceConfig{
		Scheme:  "exact",
		Network: "eip155:84532",
		PayTo:   evmAddress,
		Price:   "$0.001",
	}

	weatherAccepts, err := resourceServer.BuildPaymentRequirementsFromConfig(ctx, weatherConfig)
	if err != nil {
		return fmt.Errorf("failed to build weather payment requirements: %w", err)
	}

	forecastConfig := x402.ResourceConfig{
		Scheme:  "exact",
		Network: "eip155:84532",
		PayTo:   evmAddress,
		Price:   "$0.005",
	}

	forecastAccepts, err := resourceServer.BuildPaymentRequirementsFromConfig(ctx, forecastConfig)
	if err != nil {
		return fmt.Errorf("failed to build forecast payment requirements: %w", err)
	}

	// ========================================================================
	// STEP 4: Create payment wrappers with accepts arrays
	// ========================================================================
	paymentWrapperWeather := mcp.NewPaymentWrapper(resourceServer, mcp.PaymentWrapperConfig{
		Accepts: weatherAccepts,
		Resource: &mcp.ResourceInfo{
			URL:         "mcp://tool/get_weather",
			Description: "Get weather for a city",
			MimeType:    "application/json",
		},
	})

	paymentWrapperForecast := mcp.NewPaymentWrapper(resourceServer, mcp.PaymentWrapperConfig{
		Accepts: forecastAccepts,
		Resource: &mcp.ResourceInfo{
			URL:         "mcp://tool/get_forecast",
			Description: "Get 7-day forecast",
			MimeType:    "application/json",
		},
	})

	// ========================================================================
	// STEP 5: Register tools using REAL MCP SDK NATIVE tool registration API
	// ========================================================================

	// Free tool - works exactly as before, no changes needed
	mcpServer.AddTool(&mcpsdk.Tool{
		Name:        "ping",
		Description: "A free health check tool",
		InputSchema: map[string]interface{}{"type": "object"},
	}, func(ctx context.Context, req *mcpsdk.CallToolRequest) (*mcpsdk.CallToolResult, error) {
		return &mcpsdk.CallToolResult{
			Content: []mcpsdk.Content{
				&mcpsdk.TextContent{Text: "pong"},
			},
		}, nil
	})

	// Paid tools - wrap the handler with payment wrapper
	mcpServer.AddTool(&mcpsdk.Tool{
		Name:        "get_weather",
		Description: "Get current weather for a city. Requires payment of $0.001.",
		InputSchema: map[string]interface{}{
			"type": "object",
			"properties": map[string]interface{}{
				"city": map[string]interface{}{"type": "string", "description": "The city name"},
			},
		},
	}, paymentWrapperWeather.Wrap(func(ctx context.Context, req *mcpsdk.CallToolRequest) (*mcpsdk.CallToolResult, error) {
		city := ""
		if req.Params.Arguments != nil {
			var args map[string]interface{}
			if err := json.Unmarshal(req.Params.Arguments, &args); err == nil {
				if c, ok := args["city"].(string); ok {
					city = c
				}
			}
		}
		if city == "" {
			city = "San Francisco"
		}
		weatherData := getWeatherData(city)
		weatherJSON, _ := json.MarshalIndent(weatherData, "", "  ")
		return &mcpsdk.CallToolResult{
			Content: []mcpsdk.Content{&mcpsdk.TextContent{Text: string(weatherJSON)}},
		}, nil
	}))

	mcpServer.AddTool(&mcpsdk.Tool{
		Name:        "get_forecast",
		Description: "Get 7-day weather forecast. Requires payment of $0.005.",
		InputSchema: map[string]interface{}{
			"type": "object",
			"properties": map[string]interface{}{
				"city": map[string]interface{}{"type": "string", "description": "The city name"},
			},
		},
	}, paymentWrapperForecast.Wrap(func(ctx context.Context, req *mcpsdk.CallToolRequest) (*mcpsdk.CallToolResult, error) {
		city := ""
		if req.Params.Arguments != nil {
			var args map[string]interface{}
			if err := json.Unmarshal(req.Params.Arguments, &args); err == nil {
				if c, ok := args["city"].(string); ok {
					city = c
				}
			}
		}
		if city == "" {
			city = "San Francisco"
		}
		forecast := make([]map[string]interface{}, 7)
		for i := 0; i < 7; i++ {
			dayData := getWeatherData(city)
			dayData["day"] = i + 1
			forecast[i] = dayData
		}
		forecastJSON, _ := json.MarshalIndent(forecast, "", "  ")
		return &mcpsdk.CallToolResult{
			Content: []mcpsdk.Content{&mcpsdk.TextContent{Text: string(forecastJSON)}},
		}, nil
	}))

	// Start HTTP server with SSE transport
	return startHTTPServerExisting(mcpServer, port)
}

func startHTTPServerExisting(mcpServer *mcpsdk.Server, port string) error {
	sseHandler := mcpsdk.NewSSEHandler(func(req *http.Request) *mcpsdk.Server {
		return mcpServer
	}, nil)

	mux := http.NewServeMux()
	mux.Handle("/sse", sseHandler)
	mux.Handle("/messages", sseHandler)

	mux.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(map[string]interface{}{
			"status": "ok",
			"mode":   "existing-server",
			"server": "x402 MCP Server (Existing)",
		})
	})

	fmt.Printf("🚀 Existing MCP Server with x402 running on http://localhost:%s\n", port)
	fmt.Println("\n📋 Available tools:")
	fmt.Println("   - get_weather (paid: $0.001)")
	fmt.Println("   - get_forecast (paid: $0.005)")
	fmt.Println("   - ping (free)")
	fmt.Printf("\n🔗 Connect via SSE: http://localhost:%s/sse\n", port)
	fmt.Println("\n💡 This example shows how to add x402 to an EXISTING MCP server")
	fmt.Println("   using NewPaymentWrapper() with REAL MCP SDK.\n")

	server := &http.Server{
		Addr:    ":" + port,
		Handler: mux,
	}

	return server.ListenAndServe()
}

File: examples/typescript/servers/mcp/existing-server.ts
/**
 * MCP Server with x402 Paid Tools - Existing Server Integration
 *
 * This example demonstrates the LOW-LEVEL API using `createPaymentWrapper`.
 * Use this approach when you have an EXISTING MCP server and want to add
 * x402 payment to specific tools without adopting the full x402MCPServer abstraction.
 *
 * Key benefits:
 * - Works with your existing McpServer instance
 * - Uses native McpServer.tool() API - nothing new to learn
 * - Mix paid and free tools naturally
 * - Minimal code changes to add payment
 *
 * NOTE ON HOOKS:
 * The base McpServer from @modelcontextprotocol/sdk does NOT have lifecycle hooks.
 * Hooks are available on the CLIENT side (see x402MCPClient):
 * - beforePayment: Called before payment is created
 * - afterPayment: Called after payment is submitted
 * - onPaymentRequired: Called when payment is required
 *
 * For server-side middleware, implement custom Express middleware or wrap tool handlers.
 *
 * Run with: pnpm dev:existing
 */

import { config } from "dotenv";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { SSEServerTransport } from "@modelcontextprotocol/sdk/server/sse.js";
import { ExactEvmScheme } from "@x402/evm/exact/server";
import { createPaymentWrapper, x402ResourceServer } from "@x402/mcp";
import { HTTPFacilitatorClient } from "@x402/core/server";
import express from "express";
import { z } from "zod";

config();

const evmAddress = process.env.EVM_ADDRESS as `0x${string}`;
if (!evmAddress) {
  console.error("❌ EVM_ADDRESS environment variable is required");
  process.exit(1);
}

const facilitatorUrl = process.env.FACILITATOR_URL;
if (!facilitatorUrl) {
  console.error("❌ FACILITATOR_URL environment variable is required");
  process.exit(1);
}

const port = parseInt(process.env.PORT || "4022", 10);

/**
 * Simulates fetching weather data for a city.
 *
 * @param city - The name of the city to get weather data for
 * @returns An object containing the city name, weather condition, and temperature
 */
function getWeatherData(city: string): { city: string; weather: string; temperature: number } {
  const conditions = ["sunny", "cloudy", "rainy", "snowy", "windy"];
  const weather = conditions[Math.floor(Math.random() * conditions.length)];
  const temperature = Math.floor(Math.random() * 40) + 40;
  return { city, weather, temperature };
}

/**
 * Main entry point - Demonstrates adding x402 to an existing MCP server.
 */
export async function main(): Promise<void> {
  console.log("\n📦 Using LOW-LEVEL API (createPaymentWrapper with existing server)\n");

  // ========================================================================
  // STEP 1: Your existing MCP server (this might already exist in your code)
  // ========================================================================
  const mcpServer = new McpServer({
    name: "Existing Weather Service",
    version: "1.0.0",
  });

  // ========================================================================
  // NOTE: The base McpServer from @modelcontextprotocol/sdk does NOT have
  // lifecycle hooks (onRequest, onResponse, etc.).
  //
  // Hooks are available on the CLIENT side via x402MCPClient:
  // - beforePayment: Called before payment is created
  // - afterPayment: Called after payment is submitted
  // - onPaymentRequired: Called when payment is required
  //
  // For server-side hooks, you can wrap the tool handlers or use middleware
  // in your transport layer (Express, etc.)
  // ========================================================================

  // ========================================================================
  // STEP 2: Set up x402 resource server for payment handling
  // ========================================================================
  const facilitatorClient = new HTTPFacilitatorClient({ url: facilitatorUrl });
  const resourceServer = new x402ResourceServer(facilitatorClient);
  resourceServer.register("eip155:84532", new ExactEvmScheme());
  await resourceServer.initialize();

  // ========================================================================
  // STEP 3: Build payment requirements for different tools
  // ========================================================================
  const weatherAccepts = await resourceServer.buildPaymentRequirements({
    scheme: "exact",
    network: "eip155:84532",
    payTo: evmAddress,
    price: "$0.001",
    extra: { name: "USDC", version: "2" }, // EIP-712 domain parameters
  });

  const forecastAccepts = await resourceServer.buildPaymentRequirements({
    scheme: "exact",
    network: "eip155:84532",
    payTo: evmAddress,
    price: "$0.005",
    extra: { name: "USDC", version: "2" }, // EIP-712 domain parameters
  });

  // ========================================================================
  // STEP 4: Create payment wrappers with accepts arrays
  // ========================================================================
  const paidWeather = createPaymentWrapper(resourceServer, {
    accepts: weatherAccepts,
  });

  const paidForecast = createPaymentWrapper(resourceServer, {
    accepts: forecastAccepts,
  });

  // ========================================================================
  // STEP 5: Register tools using NATIVE McpServer.tool() API
  // ========================================================================

  // Free tool - works exactly as before, no changes needed
  mcpServer.tool("ping", "A free health check tool", {}, async () => ({
    content: [{ type: "text", text: "pong" }],
  }));

  // Paid tools - wrap the handler with payment wrapper
  // Each wrapper has its own price configured in the accepts array
  mcpServer.tool(
    "get_weather",
    "Get current weather for a city. Requires payment of $0.001.",
    { city: z.string().describe("The city name to get weather for") },
    paidWeather(async (args: { city: string }) => ({
      content: [
        {
          type: "text" as const,
          text: JSON.stringify(getWeatherData(args.city), null, 2),
        },
      ],
    })),
  );

  mcpServer.tool(
    "get_forecast",
    "Get 7-day weather forecast. Requires payment of $0.005.",
    { city: z.string().describe("The city name to get forecast for") },
    paidForecast(async (args: { city: string }) => {
      const forecast = Array.from({ length: 7 }, (_, i) => ({
        day: i + 1,
        ...getWeatherData(args.city),
      }));
      return {
        content: [{ type: "text" as const, text: JSON.stringify(forecast, null, 2) }],
      };
    }),
  );

  // Start Express server
  startExpressServer(mcpServer, port);
}

/**
 * Helper to start Express SSE server
 *
 * @param mcpServer - The MCP server instance to connect to the Express server
 * @param port - The port number to listen on
 */
function startExpressServer(mcpServer: McpServer, port: number): void {
  const app = express();
  const transports = new Map<string, SSEServerTransport>();

  app.get("/sse", async (req, res) => {
    console.log("📡 New SSE connection");
    const transport = new SSEServerTransport("/messages", res);
    const sessionId = crypto.randomUUID();
    transports.set(sessionId, transport);
    res.on("close", () => {
      console.log("📡 SSE connection closed");
      transports.delete(sessionId);
    });
    await mcpServer.connect(transport);
  });

  app.post("/messages", express.json(), async (req, res) => {
    const transport = Array.from(transports.values())[0];
    if (!transport) {
      res.status(400).json({ error: "No active SSE connection" });
      return;
    }
    await transport.handlePostMessage(req, res, req.body);
  });

  app.get("/health", (_, res) => {
    res.json({
      status: "ok",
      mode: "existing-server",
      tools: ["get_weather (paid: $0.001)", "get_forecast (paid: $0.005)", "ping (free)"],
    });
  });

  app.listen(port, () => {
    console.log(`🚀 Existing MCP Server with x402 running on http://localhost:${port}`);
    console.log(`\n📋 Available tools:`);
    console.log(`   - get_weather (paid: $0.001)`);
    console.log(`   - get_forecast (paid: $0.005)`);
    console.log(`   - ping (free)`);
    console.log(`\n🔗 Connect via SSE: http://localhost:${port}/sse`);
    console.log(`\n💡 This example shows how to add x402 to an EXISTING MCP server`);
    console.log(`   using the low-level createPaymentWrapper() API.\n`);
  });
}
